File: mfiles/gather_coord.m
function [x,y,z] = gather_coord(coordinfo,varargin)

% gather coordinates of the snapshot
% Author:   Yuanhang Huo
% Email:    yhhuo@mail.ustc.edu.cn
% Date:     2021.05.31

% ChangeLog:
% 20210814: read both cart and curv coord, --zw

%-- flags --
nargs=nargin-1;
n=1;
while n<=nargs
    
    if numel(varargin{n})==1 | ~isnumeric(varargin{n})
        switch varargin{n}
            case 'coorddir'
                coord_dir=varargin{n+1}; n=n+1;
        end
    end
    
    n=n+1;
    
end

% check
if ~ exist(coord_dir,'dir')
    error([mfilename ': directory ' coord_dir ' does not exist']);
end

% load
coordprefix='coord';
nthd=length(coordinfo);
for n=1:nthd
    
    n_i=coordinfo(n).thisid(1); n_j=coordinfo(n).thisid(2);
    i1=coordinfo(n).indxs(1); j1=coordinfo(n).indxs(2); k1=coordinfo(n).indxs(3);
    i2=coordinfo(n).indxe(1); j2=coordinfo(n).indxe(2); k2=coordinfo(n).indxe(3);
    subs=coordinfo(n).wsubs;
    subc=coordinfo(n).wsubc;
    subt=coordinfo(n).wsubt;
    fnm_coord=[coord_dir,'/',coordprefix,'_px',num2str(n_i),'_py',num2str(n_j),'.nc'];
    
    if ~ exist(fnm_coord,'file')
       error([mfilename ': file ' fnm_coord 'does not exist']);
    end

    subs=fliplr(subs);subc=fliplr(subc);subt=fliplr(subt);

    % check dimension size of x,y,z to detmine cart or curv
    xvar_info = ncinfo(fnm_coord,'x');
    num_dim_x = length(xvar_info.Dimensions);
    if num_dim_x == 1 % cart grid
      x1d = nc_varget(fnm_coord,'x',subs(3)-1,subc(3),subt(3));
      x3d = repmat(x1d,1,(j2-j1+1),(k2-k1+1));
      x(k1:k2,j1:j2,i1:i2) = permute(x3d,[3 2 1]);
    else % curv grid
      x(k1:k2,j1:j2,i1:i2)=nc_varget(fnm_coord,'x',subs-1,subc,subt);
    end

    %- y coord
    yvar_info = ncinfo(fnm_coord,'y');
    num_dim_y = length(yvar_info.Dimensions);
    if num_dim_y == 1 % cart grid
      y1d = nc_varget(fnm_coord,'y',subs(2)-1,subc(2),subt(2));
      y3d = repmat(y1d,1,(i2-i1+1),(k2-k1+1)); % y1d is a col, can't be rep as 1st dim
      y(k1:k2,j1:j2,i1:i2) = permute(y3d,[3 1 2]);
    else % curv grid
      y(k1:k2,j1:j2,i1:i2)=nc_varget(fnm_coord,'y',subs-1,subc,subt);
    end

    %- z coord
    zvar_info = ncinfo(fnm_coord,'z');
    num_dim_z = length(zvar_info.Dimensions);
    if num_dim_z == 1 % cart grid
      z1d = nc_varget(fnm_coord,'z',subs(1)-1,subc(1),subt(1));
      z3d = repmat(z1d,1,(j2-j1+1),(i2-i1+1));
      %z(k1:k2,j1:j2,i1:i2) = permute(z3d,[1 2 3]);
      z(k1:k2,j1:j2,i1:i2) = z3d;
    else % curv grid
      z(k1:k2,j1:j2,i1:i2)=nc_varget(fnm_coord,'z',subs-1,subc,subt);
    end

    % only for curv
    %x(k1:k2,j1:j2,i1:i2)=nc_varget(fnm_coord,'x',subs-1,subc,subt);
    %y(k1:k2,j1:j2,i1:i2)=nc_varget(fnm_coord,'y',subs-1,subc,subt);
    %z(k1:k2,j1:j2,i1:i2)=nc_varget(fnm_coord,'z',subs-1,subc,subt);
    
end

x=double(permute(x,[3 2 1]));
y=double(permute(y,[3 2 1]));
z=double(permute(z,[3 2 1]));


end

File: pyfiles/v1/gather_coord.py
import numpy as np
import os
from netCDF4 import Dataset

def gather_coord(coordinfo,coorddirvar,coord_dir):

    '''
    gather coordinates data
    Author: Yuanhang Huo
    Email:  yhhuo@mail.ustc.edu.cn
    Date:   2021.06.23
    '''

    if os.path.exists(coord_dir):

        coordprefix='coord'
        nthd=len(coordinfo)

        for n in range(nthd):

            n_i=coordinfo[n]['thisid'][0]
            n_j=coordinfo[n]['thisid'][1]

            i1=coordinfo[n]['indxs'][0]
            j1=coordinfo[n]['indxs'][1]
            k1=coordinfo[n]['indxs'][2]
            
            i2=coordinfo[n]['indxe'][0]
            j2=coordinfo[n]['indxe'][1]
            k2=coordinfo[n]['indxe'][2]

            subs=coordinfo[n]['wsubs']
            subc=coordinfo[n]['wsubc']
            subt=coordinfo[n]['wsubt']

            fnm_coord=coord_dir + "/" + coordprefix  + "_px" + str(n_i) + \
                     "_py" + str(n_j) + ".nc"

            if os.path.exists(fnm_coord):

                coordnc=Dataset(fnm_coord)

                subs=subs[::-1]
                subc=subc[::-1]
                subt=subt[::-1]

                # check dimension size of x,y,z to determine cart or curv
                # x
                num_dim_x=len(coordnc.variables['x'].dimensions)
                if num_dim_x == 1:
                    x1d=np.array(coordnc.variables['x'][\
                            subs[2]-1:subs[2]+subc[2]*subt[2]-1:subt[2]],\
                            dtype=np.float64)
                    x3d=np.tile(x1d,(1,j2-j1+1,k2-k1+1))
                    xblk=x3d.transpose(2,1,0)
                else:
                    xblk=np.array(coordnc.variables['x'][\
                            subs[0]-1:subs[0]+subc[0]*subt[0]-1:subt[0],\
                            subs[1]-1:subs[1]+subc[1]*subt[1]-1:subt[1],\
                            subs[2]-1:subs[2]+subc[2]*subt[2]-1:subt[2]],\
                            dtype=np.float64)

                # y
                num_dim_y=len(coordnc.variables['y'].dimensions)
                if num_dim_y == 1:
                    y1d=np.array(coordnc.variables['y'][\
                            subs[1]-1:subs[1]+subc[1]*subt[1]-1:subt[1]],\
                            dtype=np.float64)
                    y3d=np.tile(y1d,(1,i2-i1+1,k2-k1+1))
                    yblk=y3d.transpose(2,0,1)
                else:
                    yblk=np.array(coordnc.variables['y'][\
                            subs[0]-1:subs[0]+subc[0]*subt[0]-1:subt[0],\
                            subs[1]-1:subs[1]+subc[1]*subt[1]-1:subt[1],\
                            subs[2]-1:subs[2]+subc[2]*subt[2]-1:subt[2]],\
                            dtype=np.float64)

                # z
                num_dim_z=len(coordnc.variables['z'].dimensions)
                if num_dim_z == 1:
                    z1d=np.array(coordnc.variables['z'][\
                            subs[0]-1:subs[0]+subc[0]*subt[0]-1:subt[0]],\
                            dtype=np.float64)
                    z3d=np.tile(z1d,(1,j2-j1+1,i2-i1+1))
                    zblk=z3d
                else:
                    zblk=np.array(coordnc.variables['z'][\
                            subs[0]-1:subs[0]+subc[0]*subt[0]-1:subt[0],\
                            subs[1]-1:subs[1]+subc[1]*subt[1]-1:subt[1],\
                            subs[2]-1:subs[2]+subc[2]*subt[2]-1:subt[2]],\
                            dtype=np.float64)

                if n == 0:
                    xold=xblk
                    yold=yblk
                    zold=zblk

                xnew=np.empty((max(k2,xold.shape[0]),max(j2,xold.shape[1]),max(i2,xold.shape[2])))
                ynew=np.empty((max(k2,yold.shape[0]),max(j2,yold.shape[1]),max(i2,yold.shape[2])))
                znew=np.empty((max(k2,zold.shape[0]),max(j2,zold.shape[1]),max(i2,zold.shape[2])))

                xnew[:xold.shape[0],:xold.shape[1],:xold.shape[2]]=xold
                ynew[:yold.shape[0],:yold.shape[1],:yold.shape[2]]=yold
                znew[:zold.shape[0],:zold.shape[1],:zold.shape[2]]=zold

                xnew[k1-1:k2,j1-1:j2,i1-1:i2]=xblk
                ynew[k1-1:k2,j1-1:j2,i1-1:i2]=yblk
                znew[k1-1:k2,j1-1:j2,i1-1:i2]=zblk

                xold=xnew
                yold=ynew
                zold=znew

            else:
                print("Error in 'gather_coord': File " + "'" + fnm_coord + "'" + \
                      " does NOT exist.")

        xnew=xnew.transpose(2,1,0)
        ynew=ynew.transpose(2,1,0)
        znew=znew.transpose(2,1,0)

        return(xnew,ynew,znew)

    else:
        print("Error in 'gather_coord': Directory " + "'" + coord_dir + "'" + \
              " does NOT exist.")

